### PR TITLE
NDArray Slicing Clamp Bounds

### PR DESCRIPTION
**Pull request by johnc1231**
When slicing an ndarray, if someone asks for a slice bigger than the array, we need to clamp the edges of that slice. 

I will make a follow up PR that elevates `clamp` to a proper Hail function like `min` and `max`, as I think it's generally useful.

### CODE REVIEW

File: hail/src/main/scala/is/hail/asm4s/Code.scala [hunk 1]
```diff
@@ -824,6 +824,12 @@ class CodeLong(val lhs: Code[Long]) extends AnyVal {
 
   def compare(rhs: Code[Long]): Code[Int] = Code(lhs, rhs, lir.insn2(LCMP))
 
+  def max(rhs: Code[Long]): Code[Long] =
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/asm4s/Code.scala:827]: These changes snuck in an earlier approach to fix this is in Scala before I settled on clamping in python. I can remove if you'd like

File: hail/python/hail/expr/expressions/typed_expressions.py [hunk 2]
```diff
@@ -3578,15 +3578,23 @@ def __getitem__(self, item):
...
         if n_sliced_dims > 0:
             slices = []
             for i, s in enumerate(item):
                 if isinstance(s, slice):
                     step = hl.case().when(s.step != 0, s.step).or_error("Slice step cannot be zero") if s.step is not None else to_expr(1, tint64)
                     start = hl.cond(s.start >= 0, s.start, self.shape[i] + s.start) if s.start is not None else hl.cond(step >= 0, to_expr(0, tint64), self.shape[i] - 1)
+                    clamped_start = clamp(start, 0, self.shape[i] - 1)
                     stop = hl.cond(s.stop >= 0, s.stop, self.shape[i] + s.stop) if s.stop is not None else hl.cond(step >= 0, self.shape[i], to_expr(-1, tint64))
-                    slices.append(hl.tuple((start, stop, step)))
+                    clamped_stop = clamp(stop, -1, self.shape[i])
```
Comment (akotlar) [inline hail/python/hail/expr/expressions/typed_expressions.py:3596]: should t min here be -self.shape[I]?

Comment (johnc1231) [reply hail/python/hail/expr/expressions/typed_expressions.py:3596]: I don't think so. The long `stop = hl.cond(....` line above should translate everything into positive numbers.

Comment (johnc1231) [reply hail/python/hail/expr/expressions/typed_expressions.py:3596]: I only handle positive starts and stops in the emitter, this cleans them up so we match numpy semantics with indexing from the end using negatives

**Comment by johnc1231** [comment]
Suddenly getting really surprising error here about low level codegen stuff.

File: hail/src/main/scala/is/hail/asm4s/Code.scala (hunk 1 above)
Comment (akotlar) [reply hail/src/main/scala/is/hail/asm4s/Code.scala:827]: Let's remove this since not used, unneeded

**Comment by akotlar** [comment]
I see that as well, isOpenEnded.

**Pull request closed without merging**